### PR TITLE
Add Fwd tracks to StMuDST

### PR DESCRIPTION
**Pull request by jdbrice**
- Adds StMuFwdTrack to MuDST, 
- minor updates to StFwdTrackMaker (control material effects + debug tools/ttree) 
- Add StRoot/StFwdUtils : contains example fwd analysis maker

### CODE REVIEW

**Comment by jdbrice** [comment]
Testing (generate MuDSt with StMuFwdTrackCollection) can be done with:

```
setup 64b
starver dev
starver dev config/v0.2.3-rhel7-root5.34.38-64b
module load zlib-1.2.12

ln -s  StRoot/StFwdTrackMaker/macro/sim/ sim
./sim/run_batch_fast TEST 10 true >& TEST.LOG
```

this will create `sim.MuDST.root` with a StMuFwdTrack Collection filled

**Comment by jdbrice** [comment]
I see that my base branch did not include the RHICf changes to MuDst. I'll fix this and push a new commit - the should fix also the build failures

**Comment by jdbrice** [comment]
OK with commit aef975e everything should be consistent with the main branch (RHICf updates included) - Note I found a few small inconsistencies in the RHICf updates that I fixed - probably nothing practically important but wanted to note. 

So it should be ready for review at this point. I have also tested the MuDST I/O to ensure that it works (I can read StMuFwdTracks back from disk)

File: StRoot/StMuDSTMaker/COMMON/StMuFwdTrack.cxx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+// }
+
+// StPtrVecFcsCluster& StMuFwdTrack::hcalClusters() { return mHcalClusters; }
+// const StPtrVecFcsCluster& StMuFwdTrack::hcalClusters() const { return mHcalClusters; }
+// void StMuFwdTrack::addHcalCluster(StFcsCluster* p){mHcalClusters.push_back(p);}
+// void StMuFwdTrack::sortHcalClusterByET() {
+//     std::sort(mHcalClusters.begin(), mHcalClusters.end(), [](StFcsCluster* a, StFcsCluster* b) {
+//             return b->fourMomentum().perp() < a->fourMomentum().perp();
+//         });
+// }
```
Comment (plexoos) [inline StRoot/StMuDSTMaker/COMMON/StMuFwdTrack.cxx:126]:
```suggestion
```

File: StRoot/StMuDSTMaker/COMMON/StMuFwdTrackUtil.cxx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            StMuFcsCluster * muClu = fcsClusterMapEvToMu[ hClu ];
+            muFwdTrack->addHcalCluster( muClu );
+        }
+    }
+
+    return;
+}
+void StMuFwdTrackUtil::fillFwdTrack(StFwdTrackCollection*,StMuFwdTrackCollection*){
+    return;
+}
```
Comment (plexoos) [inline StRoot/StMuDSTMaker/COMMON/StMuFwdTrackUtil.cxx]:
```suggestion
```

File: StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx [hunk 3]
```diff
@@ -349,64 +355,66 @@ int StFwdTrackMaker::Init() {
...
+                    mTreeData.Crits[(n)]; mTree->Branch(n.c_str(), &mTreeData.Crits[n]);
+                    n = name + "_StraightTrackRatio";
+                    mTreeData.Crits[(n)]; mTree->Branch(n.c_str(), &mTreeData.Crits[n]);
+                }
             }
-        }
 
-        // Three hit criteria
-        path = "TrackFinder.Iteration[0].ThreeHitSegments";
-        paths = mFwdConfig.childrenOf(path);
```
Comment (plexoos) [inline StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx]:
```suggestion
```

File: StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx [hunk 4]
```diff
@@ -1077,6 +1095,7 @@ int StFwdTrackMaker::Make() {
     mFcsPreHits.clear();
     mFcsClusters.clear();
     mFwdTracks.clear();
+
```
Comment (plexoos) [inline StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx]:
```suggestion
```

File: StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx [hunk 5]
```diff
@@ -1172,60 +1187,78 @@ int StFwdTrackMaker::Make() {
...
     fwdTrack->setDidFitConvergeFully( fitStatus->isFitConvergedFully() );
     fwdTrack->setNumberOfFailedPoints( fitStatus->getNFailedPoints() );
-    // fwdTrack->setNumberOfSeedPoints(  );
+    
     fwdTrack->setNumberOfFitPoints( track->getNumPoints() );
     fwdTrack->setChi2( fitStatus->getChi2() );
     fwdTrack->setNDF( fitStatus->getNdf() );
     fwdTrack->setPval( fitStatus->getPVal() );
 
+
```
Comment (plexoos) [inline StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx]:
```suggestion
```

File: StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx [hunk 6]
```diff
@@ -1172,60 +1187,78 @@ int StFwdTrackMaker::Make() {
...
     auto cr = track->getCardinalRep();
     // charge at first point
-    fwdTrack->setCharge( cr->getCharge(track->getFittedState(0, cr)) );
+    fwdTrack->setCharge( gtr.charge );
 
     TVector3 p = cr->getMom( track->getFittedState( 0, cr ));
-    fwdTrack->setPrimaryMomentum( StThreeVectorD( p.X(), p.Y(), p.Z() ) );
+    // fwdTrack->setPrimaryMomentum( StThreeVectorD( p.X(), p.Y(), p.Z() ) );
```
Comment (plexoos) [inline StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx]:
```suggestion
```

File: StRoot/StMuDSTMaker/COMMON/StMuFwdTrackUtil.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+class StMuFwdTrackUtil : public TObject
+{
+public:
+    StMuFwdTrackCollection* getMuFwdTrack(StFwdTrackCollection*);
+    StFwdTrackCollection*   getFwdTrack(StMuFwdTrackCollection*);
+    void               fillMuFwdTrack(StMuFwdTrackCollection*,StFwdTrackCollection*, StMuFcsUtil*);
+    void               fillFwdTrack(StFwdTrackCollection*,StMuFwdTrackCollection*);
+
+
+private:
```
Comment (plexoos) [inline StRoot/StMuDSTMaker/COMMON/StMuFwdTrackUtil.h]:
```suggestion
```

File: StRoot/StMuDSTMaker/COMMON/StMuFwdTrackUtil.cxx [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    if ( evc == nullptr ){
+        LOG_WARN << "NULL StFwdTrackCollection, cannot fill MuDST" << endm;
+    }
+
+    StMuFwdTrackCollection *muc = new StMuFwdTrackCollection();
+    return muc;
+}
+StFwdTrackCollection*   StMuFwdTrackUtil::getFwdTrack(StMuFwdTrackCollection*){
+    return new StFwdTrackCollection();
+}
```
Comment (plexoos) [inline StRoot/StMuDSTMaker/COMMON/StMuFwdTrackUtil.cxx:32]:
```suggestion
}

```

File: StRoot/StMuDSTMaker/COMMON/StMuFwdTrackUtil.cxx [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+StMuFwdTrackCollection* StMuFwdTrackUtil::getMuFwdTrack(StFwdTrackCollection* evc){
+
+    if ( evc == nullptr ){
+        LOG_WARN << "NULL StFwdTrackCollection, cannot fill MuDST" << endm;
+    }
+
+    StMuFwdTrackCollection *muc = new StMuFwdTrackCollection();
+    return muc;
+}
```
Comment (plexoos) [inline StRoot/StMuDSTMaker/COMMON/StMuFwdTrackUtil.cxx:28]:
```suggestion
}

```

File: StRoot/StMuDSTMaker/COMMON/StMuFwdTrackCollection.cxx [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ * Description: Fwd Tracks
+ *
+ ***************************************************************************/
+
+#include "StMuDSTMaker/COMMON/StMuFwdTrackCollection.h"
+#include "StMuDSTMaker/COMMON/StMuFwdTrack.h"
+
+#include "St_base/StMessMgr.h"
+
+ClassImp(StMuFwdTrackCollection)
```
Comment (plexoos) [inline StRoot/StMuDSTMaker/COMMON/StMuFwdTrackCollection.cxx]:
```suggestion
```

File: StRoot/StMuDSTMaker/COMMON/StMuFwdTrackCollection.cxx [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    return newFwdTrack;
+}
+
+unsigned int StMuFwdTrackCollection::numberOfFwdTracks() const{
+    if(!mFwdTracks) return 0;
+    return mFwdTracks->GetEntriesFast();
+}
+
+StMuFwdTrack*  StMuFwdTrackCollection::getFwdTrack(int index){
+    if(!mFwdTracks) return NULL;
```
Comment (plexoos) [inline StRoot/StMuDSTMaker/COMMON/StMuFwdTrackCollection.cxx]:
```suggestion
    if (!mFwdTracks) return nullptr;
```

File: StRoot/StFwdUtils/StFwdAnalysisMaker.cxx [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+#include "StarMagField/StarMagField.h"
+
+#include "St_base/StMessMgr.h"
+#include "StarClassLibrary/StPhysicalHelix.hh"
+#include "StarClassLibrary/SystemOfUnits.h"
+
+
+#include "TROOT.h"
+#include "TLorentzVector.h"
```
Comment (plexoos) [inline StRoot/StFwdUtils/StFwdAnalysisMaker.cxx]:
```suggestion
```

File: StRoot/StFwdUtils/StFwdAnalysisMaker.cxx [hunk 13]
```diff
@@ -0,0 +1,127 @@
+#include "StFwdUtils/StFwdAnalysisMaker.h"
+#include "StFwdTrackMaker/Common.h"
+
+#include "TMath.h"
+
```
Comment (plexoos) [inline StRoot/StFwdUtils/StFwdAnalysisMaker.cxx]:
```suggestion
```

File: StRoot/StFwdUtils/StFwdAnalysisMaker.cxx [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+#include <limits>
+#include <map>
+#include <string>
+#include <string>
+#include <vector>
+
+#include "StBFChain/StBFChain.h"
+
```
Comment (plexoos) [inline StRoot/StFwdUtils/StFwdAnalysisMaker.cxx]:
```suggestion
```

File: StRoot/StFwdUtils/StFwdAnalysisMaker.h [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+    int Init();
+    int Finish();
+    int Make();
+    void Clear(const Option_t *opts = "");
+    void ProcessFwdTracks();
+    void ProcessFwdMuTracks();
+
+  private:
+  protected:
```
Comment (plexoos) [inline StRoot/StFwdUtils/StFwdAnalysisMaker.h]:
```suggestion
```

**Pull request merged**